base/src/main/java/org/eclipse/serializer/concurrency/StripeLockedExecutor.java

```diff
@@ -0,0 +1,241 @@
+ * 
+ * @see ReentrantLock
+ * @see ReadWriteLock
+ */
+public interface StripeLockedExecutor
+{
+	/**
+	 * Executes an operation protected by a read lock.
+	 *
+	 * @param mutex the mutex to lock on, not <code>null</code>
+	 * @param action the action to execute
+	 */
+	public void read(Object mutex, Action action);
+	
+	/**
+	 * Executes an operation protected by a read lock.
+	 *
+	 * @param <R> the producer's return type
+	 * @param mutex the mutex to lock on, not <code>null</code>
+	 * @param producer the producer to execute
+	 * @return the producer's result
+	 */
+	public <R> R read(Object mutex, Producer<R> producer);
+	
+	/**
+	 * Executes an operation protected by a write lock.
+	 *
+	 * @param mutex the mutex to lock on, not <code>null</code>
+	 * @param action the action to execute
+	 */
+	public void write(Object mutex, Action action);
+	
+	/**
+	 * Executes an operation protected by a write lock.
+	 *
+	 * @param <R> the producer's return type
+	 * @param mutex the mutex to lock on, not <code>null</code>
+	 * @param producer the producer to execute
+	 * @return the producer's result
+	 */
+	public <R> R write(Object mutex, Producer<R> producer);
+	
+	
+	
+	public static final class Static
+	{
+		private final static Object LOCK = new Object();
+		private static volatile StripeLockedExecutor sharedInstance;
+		
+		public static StripeLockedExecutor sharedInstance()
+		{
+			/*
+			 * Double-checked locking to reduce the overhead of acquiring a lock
+			 * by testing the locking criterion.
+			 * The field (Static.sharedInstance) has to be volatile.
+			 */
+			StripeLockedExecutor sharedInstance = Static.sharedInstance;
+			if(sharedInstance == null)
+			{
+				synchronized(LOCK)
+				{
+					if((sharedInstance = Static.sharedInstance) == null)
+					{
+						sharedInstance = Static.sharedInstance = StripeLockedExecutor.New(
+							Runtime.getRuntime().availableProcessors()
+						);
+					}
+				}
+			}
+			return sharedInstance;
+		}
+		
+		
+		private Static()
+		{
+			// static only
+			throw new UnsupportedOperationException();
+		}
+	}
+	
+	
+	/**
+	 * Provides a global {@link StripeLockedExecutor} instance.
+	 * <p>
+	 * Only a single one exists for the whole VM process, meaning it can be used to create VM-wide locks.
+	 * 
+	 * @return a shared {@link StripeLockedExecutor} instance
+	 */
+	public static StripeLockedExecutor global()
+	{
+		return Static.sharedInstance();
+	}
+	
+	
+	
+	/**
+	 * Pseudo-constructor method to create a new {@link StripeLockedExecutor}.
+	 * 
+	 * @param stripeCount maximum number of stripes
+	 * @return a newly created {@link StripeLockedExecutor}
+	 */
+	public static StripeLockedExecutor New(final int stripeCount)
+	{
+		return new StripeLockedExecutor.Default(
+			positive(stripeCount)
+		);
+	}
+	
+	
+	public static class Default implements StripeLockedExecutor
+	{
+		private transient volatile ReentrantReadWriteLock[] reentrantLocks;
```

Comment:
field reentrantLocks may be final

Reply:
field can't be volatile and final at the same time, volatile is needed for double-checked locking